### ISSUE TITLE
Assembly redirects not honored?

### ISSUE THREAD

**Issue by dennisdoomen** [gluck/il-repack#122]
I have an assembly that references v5.2.3 of `System.Web.Http` (aka WebAPI 2) and also uses `Autofac.Integration.WebAPI` which has a reference to 5.2.0 of `System.Web.Http`. The original assembly has the following assembly redirects:
```
      <dependentAssembly>
        <assemblyIdentity name="System.Web.Http" publicKeyToken="31bf3856ad364e35" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-5.2.3.0" newVersion="5.2.3.0" />
      </dependentAssembly>
      <dependentAssembly>
```
When I run ILRepack 2.0.2, I get the following error 

```
ERROR: Failed to load assembly C:\Workspaces\OnSiteIssuer.Middleware\Sources/Middleware/bin/Release/Autofac.Integration.
WebApi.dll
Mono.Cecil.AssemblyResolutionException: Failed to resolve assembly: 'System.Web.Http, Version=5.2.0.0, Culture=neutral,
PublicKeyToken=31bf3856ad364e35'
```

Am I doing something wrong? Or are those assembly redirects not (yet) honored?

**Comment by gluck** [comment]
Hi Dennis,

AFAIK assembly redirect only apply at runtime to the exe, I don't believe they get applied to referenced dlls.
But I believe ILRepack doesn't care about the version, so should you pass it a "/lib:<FOLDER>" argument with System.Web.Http (not matter the version which is there, as long as it contains all things referenced), I believe you'll be just fine.

(Ideally ILRepack wouldn't even need to load those referenced assemblies, but for some reason Cecil needs to)

Let me know if you tried that already.

**Comment by dennisdoomen** [comment]
So are you saying that I should copy the `.config` of the source assembly and use it as the `.config` for `ilrepack.exe`?

**Comment by gluck** [comment]
No, cecil will ignore it anyway, the .config is used by the CLR assembly loader, not by cecil.
I was suggesting you add "/lib:C:\tmp" to the ilrepack args, with System.Web.Http.dll (any version will probably work) inside C:\tmp

**Comment by dennisdoomen** [comment]
That seems to work, but now I'm getting a couple of other warnings that worry me. I managed to reproduce this problem in a public repo. Just clone https://github.com/dennisdoomen/piercer, check ou the `AttemptToUseILRepack` branch and run `build.bat`.

**Comment by gluck** [comment]
What happens is that `Swashbuckle.Core` has been compiled with `System.Net.Http.Formatting` 4.0.0.0 (nuget 4.0.20710) which used to contain a property `SerializerSettings` on `JsonMediaTypeFormatter`.
This property has later been moved to base class `BaseJsonMediaTypeFormatter`, which is the case in the version you're merging.

When going through the IL, ILRepacks notices calls to get/set `SerializerSettings::JsonMediaTypeFormatter` which it can't resolve.
(done in `Swashbuckle.Application.HttpConfigurationExtensions` and `Swashbuckle.Application.SwaggerDocsHandler`).

It (suprisingly) merges it ok, and your merged assembly no longer have any reference to `System.Net.Http.Formatting`, but I'm not 100% the code will work (with or without repack, doesn't matter as the repacked code is the same), supposedly it shouldn't as the property isn't virtual, the the IL code is a `callvirt`, which (maybe) will resolve the base property just fine.

**Comment by gluck** [comment]
(also the resulting assembly goes fine through peverify, which means it's probably ok)

TL;DR ILRepack works its way through the process (albeit with this warning, which  should be removed in that case), but only because the moved property/methods are used only in `callvirt` context. Should they be used in `call` ones (e.g. by overriding and calling `base.method()`), it'd fail.

**Comment by dennisdoomen** [comment]
I don't get that. If `Shashbuckler.Core` is using an older version, `ILRepack` shouldn't change that. IMHO, that's the beauty of merging assemblies. Each components merges as much of its internal dependencies so that the components that use yours, don't have to think about that.

**Comment by gluck** [comment]
Oh, you expect ILRepack to merge `System.Net.Http.Formatting` 4.0.0.0 in the resulting assembly ?
Right now this assembly isn't even referenced, so you'd have to provide it somehow.

Note also that it will only work if `Shashbuckler.Core` only uses this assembly internally, and doesn't expose types from it (e.g. as method return types or params). In that case having multiple versions of an assembly will fail at runtime.

Merging assemblies is no magic, it's usually referred as "static-linking", and is only a way to "finalize/set-in-stone" a set of assemblies, so that they aren't subject to dynamic assembly loading (hell) anymore.

**Comment by dennisdoomen** [comment]
Oh wait, I thought `Swashbuckle.Core` internalized `System.Net.Http.Formatting`. So we're essentially talking about incompatible dependencies, aren't we?

**Comment by gluck** [comment]
We are, except with a bit of luck (and some limitations of the usage), they are pretty much compatible, at least from the PoV of this property moved.
That being said, there's only so much which can be staticly verified, so you're indeed taking some risk with such a gap (if following semver, you can't really expect 4.0 and 5.0 to be compatible).

**Comment by dennisdoomen** [comment]
Agreed. But just to be clear, if Swashbuckle would have internalized that assembly, everything would be fine and dandy, wouldn't it?

**Comment by gluck** [comment]
Yup, like most ppl do with Newtonsoft.Json (except the aspnet team who decided to re-implement an internal json parser instead :) )

**Comment by dennisdoomen** [comment]
Thanks. That closes this issue.

**Comment by rolandh** [comment]
Assembly redirects don't seem to be handled properly. I end up with things like this in my final app.config file.

`      <dependentAssembly>
        <assemblyIdentity name="System.Net.Http.Formatting" publicKeyToken="31bf3856ad364e35" culture="neutral" />
        <assemblyIdentity name="protobuf-net" publicKeyToken="257B51D87D2E4D67" culture="neutral" />
        <assemblyIdentity name="System.Net.Http" publicKeyToken="B03F5F7F11D50A3A" culture="neutral" />
        <assemblyIdentity name="protobuf-net" publicKeyToken="257b51d87d2e4d67" culture="neutral" />
        <assemblyIdentity name="Microsoft.Extensions.Primitives" publicKeyToken="adb9793829ddae60" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-5.2.3.0" newVersion="5.2.3.0" />
        <bindingRedirect oldVersion="0.0.0.0-2.4.0.0" newVersion="2.4.0.0" />
        <bindingRedirect oldVersion="0.0.0.0-4.2.0.0" newVersion="4.0.0.0" />
        <bindingRedirect oldVersion="0.0.0.0-2.3.2.0" newVersion="2.3.2.0" />
        <bindingRedirect oldVersion="0.0.0.0-2.0.0.0" newVersion="2.0.0.0" />
      </dependentAssembly>`
